src/downshift.js

```diff
@@ -287,6 +290,14 @@ class Downshift extends Component {
           onChangeArg = stateToSet.selectedItem
         }
         stateToSet.type = stateToSet.type || Downshift.stateChangeTypes.unknown
+
+        if (stateToSet.hasOwnProperty('inputValue')) {
```

Comment:
Unfortunately, this wont work with [when we call `reset`](https://github.com/paypal/downshift/blob/1385fc9dc5bc5e93ef68c863c971946a3e29db63/src/downshift.js#L621) which users can do themselves _or_ when the user clicks outside the menu... 🤔 This is because `stateToSet` in that case is a function, but it does set the `inputValue`. So I'm thinking that we can do this, but we _also_ have to handle the case where `stateToSet` is a function. I think the best thing to do would be to test this by calling `reset` and make sure that `onInputValueChange` is called in that case (in addition to the other tests you already have). Then we'll figure out the best way to implement that.

Reply:
Did you see my comment here? https://github.com/paypal/downshift/pull/229#discussion_r145216365
I'm not sure yet that invoking the `onInputValueChange` callback inside the `setState` function works at all to fix the issue at hand. I'd like to prove that first before addressing issues like the different incarnations of  `stateToSet`.

Sorry for the confusion with two opened PRs. :/

Reply:
I did see your comment. But as I said, the existing code will _never_ call `onInputValueChange` when `reset` is called. We definitely need to call `onInputValueChange` any time the `inputValue` is called.

I think that the solution is good as is, then we add _more_ code for the case when the `stateToSet` is a function (which only applies in the `reset` case). Luckily for us, when `reset` happens, the original issue won't be a problem.

Reply:
The issue I'm having is that calling `onInputValueChange` inside `setState` function does not appear to fix the issue with the cursor jumping to the end of the input. 

Or am I misunderstanding and with "the existing code" you were referring to my original version, where the `onInputValueChange` call was placed above the `setState`?

Sorry, I'm a bit confused right now.

Reply:
Ok, so the code you have already is great. We're not going to change it.

The problem is that the code you have doesn't cover the `reset` case because `stateToSet` is a function in that case (and it has to be that way). So if we merged this PR as is it would work great for everything except for when the user clicks outside the menu or tabs out of the menu (which is when `reset` is called).

To deal with this issue, we need more code. We need to handle the case where `stateToSet` is a function. If it is, then we need to do that inside of `setState`.

Luckily, when `stateToSet` is a function, then we wont have the problem of the cursor jumping to the end of the input because the focus will be elsewhere anyway.

Does that make sense?

Reply:
Yes, it does. Thanks for clearing things up! :)